File: app/src/main/java/com/giaquino/android/cropped/data/common/Resource.java
package com.giaquino.android.cropped.data.common;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.util.Consumer;

public class Resource<T> {

    public enum Status {
        SUCCESS, FAILURE, LOADING
    }

    public static <T> Resource<T> success(@NonNull T data) {
        return new Resource<>(data, null, Status.SUCCESS);
    }

    public static <T> Resource<T> failure(@NonNull Throwable error) {
        return new Resource<>(null, error, Status.FAILURE);
    }

    public static <T> Resource<T> loading(@Nullable T data) {
        return new Resource<>(data, null, Status.LOADING);
    }

    @Nullable
    public final T data;

    @Nullable
    public final Throwable error;

    @NonNull
    public final Status status;

    private Resource(@Nullable T data, @Nullable Throwable error, @NonNull Status status) {
        this.data = data;
        this.error = error;
        this.status = status;
    }

    public void isSuccess(Consumer<T> consumer) {
        if (status == Status.SUCCESS) consumer.accept(data);
    }

    public void isFailure(Consumer<Throwable> consumer) {
        if (status == Status.FAILURE) consumer.accept(error);
    }

    public void isLoading(Consumer<T> consumer) {
        if (status == Status.LOADING) consumer.accept(data);
    }
}

File: app/src/main/java/com/giaquino/android/cropped/data/common/Resource.kt
package com.giaquino.android.cropped.data.common

class Resource<T> private constructor(
        private val data: T?,
        private val error: Throwable?,
        private val status: Status) {

    enum class Status {
        SUCCESS, FAILURE, LOADING
    }

    fun isSuccess(accept: (T) -> Unit) {
        if (status == Status.SUCCESS) accept(data!!)
    }

    fun isFailure(accept: (Throwable) -> Unit) {
        if (status == Status.FAILURE) accept(error!!)
    }

    fun isLoading(accept: (T?) -> Unit) {
        if (status == Status.LOADING) accept(data)
    }

    companion object {
        fun <T> success(data: T): Resource<T> {
            return Resource(data, null, Status.SUCCESS)
        }

        fun <T> failure(error: Throwable): Resource<T> {
            return Resource(null, error, Status.FAILURE)
        }

        fun <T> loading(data: T?): Resource<T> {
            return Resource(data, null, Status.LOADING)
        }
    }
}